### PR TITLE
Add dra, ddec, and distance to PrecoveryCandidate

### LINKED ISSUE #5: Add dra, ddec, and distance to PrecoveryCandidate
Additional useful quantities to include in the [PrecoveryCandidate](https://github.com/B612-Asteroid-Institute/precovery/blob/master/precovery/precovery_db.py#L18-L25) dataclass are the sky-plane residuals and the great circle distance.

### PR DESCRIPTION
**Pull request by moeyensj**
Work to address #5. This PR adds dra, ddec, and distance.

### CODE REVIEW

File: precovery/precovery_db.py [hunk 1]
```diff
@@ -207,14 +210,18 @@ def _check_frames(
             n = 0
             for obs in self.frames.iterate_observations(f):
                 n += 1
-                if obs.matches(exact_ephem, tolerance):
+                matches, dra, ddec, distance = obs.matches(exact_ephem, tolerance)
```
Comment (moeyensj) [inline precovery/precovery_db.py]: Would it be cleaner if we redefine `obs.matches` to `obs.distance` and have this function return `dra, ddec, distance`? We can then delegate the check against `tolerance` in this for-loop rather than returning the boolean.

**Comment by moeyensj** [comment]
@spenczar Could you take a look at the changes to the `obs.matches` function and let me know your thoughts? There are two other commits in this PR, one that fixes a timedelta bug and another that disables the linear propagator. Disabling the linear propagator is a temporary change while we push forward on getting some of the candidates longer arcs of observation. Falling back to 2-body as our "approximate" propagation has made it so this code agrees 1-1 with the observations from brute-force attribution.

File: precovery/precovery_db.py (hunk 1 above)
Comment (spenczar) [reply precovery/precovery_db.py]: Yes, definitely. `matches` should return just a `bool`, based on the name. `distance` seems like a good name for a method that computes distances.

File: precovery/frame_db.py [hunk 2]
```diff
@@ -174,6 +176,11 @@ def get_frames(
     ) -> Iterator[HealpixFrame]:
         """
         Yield all the frames which are for given obscode, MJD, healpix.
+
+        MJDs are checked to within +- 1e-7 days or 8.64 ms. Any frames that
+        are within 8.64 ms of the given mjd will be returned. This does not garauntee
+        that they will represent the desired exposure time and may lead to multiple
+        matches computed at the wrong observation time.
```
Comment (spenczar) [inline precovery/frame_db.py:190]: Oof, this is annoying, isn't it?

This is just a problem with floating point timestamps, I suppose. Perhaps there is a way we can log at the WARN level if we get multiple hits?

File: precovery/orbit.py [hunk 3]
```diff
@@ -249,7 +249,7 @@ def approximately_propagate(
 
         Accuracy will decrease as timedelta increases.
         """
-        if timedelta <= 3.0 and self.ra_velocity < 1 and self.dec_velocity < 1:
+        if np.abs(timedelta) <= -1.0 and self.ra_velocity < 1 and self.dec_velocity < 1:
```
Comment (spenczar) [inline precovery/orbit.py:252]: Good catch to use `np.abs` - but why the change of the RHS to `<= -1`? It seems like that expression will always be `false` now.

**Review by spenczar — CHANGES_REQUESTED** [review]
Some good catches here, but I think moving the distance calculations out of `matches` and into a new `distance` method seems much better.

File: precovery/orbit.py (hunk 3 above)
Comment (moeyensj) [reply precovery/orbit.py:252]: This is just a temporary change to disable the approximate propagator. Disabling the approximate propagation helped recover missing observations for some of the objects we were looking at. The more sophisticated approach would be to analyze the trade-off between healpixel size, approximate propagation time scale, and the predicted RA and Dec velocites. For the time being, however, using 2-body by default and then N-body propagations for the more precise propagations is thankfully fast enough.

File: precovery/precovery_db.py (hunk 1 above)
Comment (moeyensj) [reply precovery/precovery_db.py]: Done, I've changed the matches function to distance and move the check against tolerance to the for-loop in [f4ad4d5](https://github.com/B612-Asteroid-Institute/precovery/pull/8/commits/f4ad4d53b5363eefa9b51ba681a4795edb5f481b).

File: precovery/frame_db.py (hunk 2 above)
Comment (moeyensj) [reply precovery/frame_db.py:190]: I'd suggest we reduce the 1e-7 further to 1e-12 or 0.0864 microseconds (what is the smallest comparison we could go to numerically?). We should hopefully never (🤞) encounter two exposures within 0.0864 microseconds from the same observatory. 

Checking for MJD uniqueness would require (I think) looping over the rows first before yielding them back and tracking the unique number of MJDs. It's not fully obvious how much more expensive that would be but it is something that is totally doable. Something as simple as this could work? 
```
rows = self.dbconn.execute(select_stmt)

mjds = set()
for r in rows:
    mjds.add(r[3])

if len(mjds) > 1:
   logger.warn(f"Query returned non-unique MJDs for mjd: {mjd}, healpix: {int(healpixel)}, obscode: {obscode}.")
 
for r in rows:
    yield HealpixFrame(*r)
```

File: precovery/precovery_db.py (hunk 1 above)
Comment (moeyensj) [reply precovery/precovery_db.py]: Done, I've changed the matches function to distance and moved the check against tolerance to the for-loop in [f4ad4d5](https://github.com/B612-Asteroid-Institute/precovery/pull/8/commits/f4ad4d53b5363eefa9b51ba681a4795edb5f481b).

File: precovery/frame_db.py (hunk 2 above)
Comment (spenczar) [reply precovery/frame_db.py:190]: This is really fraught with peril. Floating point inaccuracies are really hard to predict. See https://www.sqlite.org/floatingpoint.html - a serious applied mathematician could probably put error bounds on this but that's way beyond my abilities.

Your loop is valid and fast, and the warning is probably sufficient. I think that's very preferable over shrinking down to less than a microsecond.

The real Right Thing here is to stop using floating point values. We should store times as integers if we want to do exact matching. That could be done by using something like UTC milliseconds, which should map cleanly from Julian Dates. Since sqlite uses 64-bit integers, this should be valid for essentially forever. Another option would be to multiply our MJDs by a constant integer factor (say, 1e9) and round to the nearest integer.

Then, we could actually do exact matches and thing should work correctly.

File: precovery/orbit.py (hunk 3 above)
Comment (spenczar) [reply precovery/orbit.py:252]: Got it. Let's just remove approximate propagation entirely, then, and use 2-body. That will mean a lot of code deletion - which is a very good thing.

Comment (moeyensj) [reply precovery/orbit.py:252]: The code is fast enough for now but once we start indexing bigger datasets we may need to go back and look for faster methods. The linear propagation should work fine for time regimes that are small (hours to days, for most objects). Let's keep it for now so we have a starting place for future optimization.

File: precovery/frame_db.py (hunk 2 above)
Comment (moeyensj) [reply precovery/frame_db.py:190]: I've added a logging statement at the WARN level in [0799cd1](https://github.com/B612-Asteroid-Institute/precovery/pull/8/commits/0799cd128f1e8da670f7c003cb35294af3f8dcd2) and kept the time range at +- 1e-7 days.

**Comment by moeyensj** [comment]
@spenczar I think this is ready for another look.

File: precovery/orbit.py (hunk 3 above)
Comment (spenczar) [reply precovery/orbit.py:252]: Okay - but I still think this could be clearer. It could be (for example) `if False and np.abs(timedelta) <= 1.0 and ...`, or at least could have a comment. The current thing just looks very odd, and future optimization might come in, like, two years, and by different people, so we should be clearer here.

**Review by spenczar — APPROVED** [review]

**Comment by moeyensj** [comment]
Closes #5.
